routing/router.go

```diff
@@ -1439,43 +1439,38 @@ func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
 		return nil, err
 	}
 
-	// If no amount is specified, we need to build a route for the minimum
-	// amount that this route can carry.
-	useMinAmt := amt == nil
-
 	var (
 		receiverAmt lnwire.MilliSatoshi
+		senderAmt   lnwire.MilliSatoshi
 		pathEdges   []*unifiedEdge
 	)
 
-	if useMinAmt {
-		// For minimum amount routes, aim to deliver at least 1 msat to
-		// the destination. There are nodes in the wild that have a
-		// min_htlc channel policy of zero, which could lead to a zero
-		// amount payment being made.
-		var senderAmt lnwire.MilliSatoshi
-		pathEdges, senderAmt, err = senderAmtBackwardPass(
-			unifiers, useMinAmt, 1, bandwidthHints,
-		)
-		if err != nil {
-			return nil, err
-		}
+	// We determine the edges compatible with the requested amount, as well
+	// as the amount to send, which can be used to determine the final
+	// receiver amount, if a minimal amount was requested.
+	pathEdges, senderAmt, err = senderAmtBackwardPass(
+		unifiers, amt, bandwidthHints,
+	)
+	if err != nil {
+		return nil, err
+	}
 
-		receiverAmt, err = receiverAmtForwardPass(senderAmt, pathEdges)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		// If an amount is specified, we need to build a route that
-		// delivers exactly this amount to the final destination.
-		pathEdges, _, err = senderAmtBackwardPass(
-			unifiers, useMinAmt, *amt, bandwidthHints,
-		)
-		if err != nil {
-			return nil, err
-		}
+	// For the minimal amount search, we need to do a forward pass to find a
+	// larger receiver amount due to possible min HTLC bumps, otherwise we
+	// just use the requested amount.
+	receiverAmt, err = fn.ElimOption(
+		amt,
+		func() fn.Result[lnwire.MilliSatoshi] {
+			receiverAmt, err := receiverAmtForwardPass(
+				senderAmt, pathEdges,
+			)
 
-		receiverAmt = *amt
+			return fn.NewResult(receiverAmt, err)
```

Comment:
I learned from @Roasbeef that you can actually just do this and it mysteriously works:
```golang
func() fn.Result[lnwire.MilliSatoshi] {
        return fn.NewResult(
                receiverAmtForwardPass(
                        senderAmt, pathEdges,
                )
        )
}
```